internal/auth/providers/amazon_cognito.go

```diff
@@ -0,0 +1,395 @@
+	}
+
+	return userinfo.EmailAddress, nil
+}
+
+// ValidateGroupMembership takes in an email, a list of allowed groups an access token
+// - compares allowed groups to users' groups.
+// Conditionally returns an empty (nil) slice or a slice of the matching groups.
+func (p *AmazonCognitoProvider) ValidateGroupMembership(email string, allowedGroups []string, accessToken string) ([]string, error) {
+	// TODO: This is currently non-functioning
+	return allowedGroups, nil
+}
+
+// GetUserProfile takes in an access token and sends a request to the /userinfo endpoint.
+// Conditionally returns nil response or a struct of specified claims.
+func (p *AmazonCognitoProvider) GetUserProfile(AccessToken string) (*GetUserProfileResponse, error) {
+	response := &GetUserProfileResponse{}
+
+	if AccessToken == "" {
+		return nil, ErrBadRequest
+	}
+
+	bearer := fmt.Sprintf("Bearer %s", AccessToken)
+	header := http.Header{}
+	header.Set("Authorization", bearer)
+
+	err := p.amazonCognitoRequest("GET", p.ProfileURL.String(), nil, []string{"action:userinfo"}, header, false, response)
+	if err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+// RefreshSessionIfNeeded takes in a SessionState and
+// returns false if the session is not refreshed and true if it is.
+func (p *AmazonCognitoProvider) RefreshSessionIfNeeded(s *sessions.SessionState) (bool, error) {
+	if s == nil || !s.RefreshPeriodExpired() || s.RefreshToken == "" {
+		return false, nil
+	}
+
+	newToken, duration, err := p.RefreshAccessToken(s.RefreshToken)
+	if err != nil {
+		return false, err
+	}
+	logger := log.NewLogEntry()
+
+	s.AccessToken = newToken
+
+	s.RefreshDeadline = time.Now().Add(duration).Truncate(time.Second)
+	logger.WithUser(s.Email).WithRefreshDeadline(s.RefreshDeadline).Info("refreshed access token")
+
+	return true, nil
+}
+
+// RefreshAccessToken takes in a refresh token and returns the new access token along with an expiration date.
+func (p *AmazonCognitoProvider) RefreshAccessToken(refreshToken string) (token string, expires time.Duration, err error) {
+	params := url.Values{}
+	params.Add("client_id", p.ClientID)
+	params.Add("refresh_token", refreshToken)
+	params.Add("grant_type", "refresh_token")
+
+	var response struct {
+		AccessToken string `json:"access_token"`
+		ExpiresIn   int64  `json:"expires_in"`
+	}
+
+	err = p.amazonCognitoRequest("POST", p.RedeemURL.String(), params, []string{"action:refresh"}, nil, true, &response)
+	if err != nil {
+		return "", 0, err
+	}
+
+	return response.AccessToken, time.Duration(response.ExpiresIn) * time.Second, nil
+}
+
+// Revoke revokes the refresh token from a given session state.
+// For the Amazon Cognito provider this triggers a global sign out for this user on all devices
+func (p *AmazonCognitoProvider) Revoke(s *sessions.SessionState) error {
+	_, err := p.cip.GlobalSignOut(&cognitoidentityprovider.GlobalSignOutInput{
+		AccessToken: &s.AccessToken,
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
```

Comment:
Because this doesn't need to use the `amazonCognitoRequest` function, we lose the metric tags that are normally passed in with it. Is it worth adding those tags back in manually here?